### PR TITLE
feat: Optimize space occupation of pie charts titles

### PR DESCRIPTION
**Pull request by fralongo**
### Description

Distributes pie charts labels to better occupy available space by:
- Positioning them in a circle around the chart, as opposed to aligning them on the right/left
- Pushing them as much as possible to the top or bottom of the chart, were there is more horizontal space to display the labels

### How has this been tested?

Locally and through internal pipeline

<details>
   <summary>Review checklist</summary>

_The following items are to be evaluated by the author(s) and the reviewer(s)._

#### Correctness

- _Changes include appropriate documentation updates._
- _Changes are backward-compatible if not indicated, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#public-apis)._
- _Changes do not include unsupported browser features, see [`CONTRIBUTING.md`](https://github.com/cloudscape-design/components/blob/main/CONTRIBUTING.md#browsers-support)._
- _Changes were manually tested for accessibility, see [accessibility guidelines](https://cloudscape.design/foundation/core-principles/accessibility/)._

#### Security

- _If the code handles URLs: all URLs are validated through [the `checkSafeUrl` function](https://github.com/cloudscape-design/components/blob/main/src/internal/utils/check-safe-url.ts)._

#### Testing

- _Changes are covered with new/existing unit tests?_ Yes
- _Changes are covered with new/existing integration tests?_ No
</details>

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/pie-chart/utils.ts [hunk 1]
```diff
@@ -163,23 +164,70 @@ export const balanceLabelNodes = (
...
+      box.x += xOffset;
     }
 
     previousBBox = box;
   }
 };
+
+const squareDistance = (edge: Array<number>): number => Math.pow(edge[0], 2) + Math.pow(edge[1], 2);
+
+const computeXOffset = (box: { x: number; y: number; height: number }, yOffset: number, radius: number): number => {
```
Comment (rubencarvalho) [inline src/pie-chart/utils.ts:193]: Could we add some unit tests to this computeXOffset?

File: src/pie-chart/labels.tsx [hunk 2]
```diff
@@ -93,16 +95,16 @@ export default <T extends PieChartProps.Datum>({
...
       const labelDatum = pieData[i];
-      const midAngle = labelDatum.startAngle + (labelDatum.endAngle - labelDatum.startAngle) / 2;
+      const smartAngle = computeSmartAngle(labelDatum.startAngle, labelDatum.endAngle, optimizeLabels);
 
       // Make the marker line longer if the segment is closer to the top or bottom of the chart
-      arcMarkerBreak.outerRadius(radius + 20 * (0.5 * Math.cos(2 * midAngle) + 0.5));
-      arcMarkerBreak.innerRadius(radius + 20 * (0.5 * Math.cos(2 * midAngle) + 0.5));
-      const [startX, startY] = arcMarkerStart.centroid(datum);
-      const [breakX, breakY] = arcMarkerBreak.centroid(datum);
+      arcMarkerBreak.outerRadius(radius + 20 * (0.5 * Math.cos(2 * smartAngle) + 0.5));
```
[the remark this answers is not in the document]
Comment (fralongo) [reply src/pie-chart/labels.tsx]: Done

File: src/pie-chart/labels.tsx [hunk 3]
```diff
@@ -78,7 +78,9 @@ export default <T extends PieChartProps.Datum>({
   containerRef,
 }: LabelsProps<T>) => {
   const containerBoundaries = useElementBoundaries(containerRef);
-
+  const optimizeLabels =
```
[the remark this answers is not in the document]
Comment (fralongo) [reply src/pie-chart/labels.tsx]: Done

File: src/pie-chart/utils.ts [hunk 4]
```diff
@@ -163,23 +164,70 @@ export const balanceLabelNodes = (
...
-      box.y += offset;
+      box.y += yOffset;
+      box.x += xOffset;
     }
 
     previousBBox = box;
   }
 };
+
+const squareDistance = (edge: Array<number>): number => Math.pow(edge[0], 2) + Math.pow(edge[1], 2);
```
[the remark this answers is not in the document]
Comment (fralongo) [reply src/pie-chart/utils.ts]: Done

File: src/pie-chart/utils.ts [hunk 5]
```diff
@@ -163,23 +164,70 @@ export const balanceLabelNodes = (
...
+
+const computeXOffset = (box: { x: number; y: number; height: number }, yOffset: number, radius: number): number => {
+  const edges = [
+    [box.x, box.y + yOffset],
+    [box.x, box.y + box.height + yOffset],
+  ];
+  const closestEdge = squareDistance(edges[0]) < squareDistance(edges[1]) ? edges[0] : edges[1];
+
+  if (squareDistance(closestEdge) < Math.pow(radius, 2)) {
```
[the remark this answers is not in the document]
Comment (fralongo) [reply src/pie-chart/utils.ts:199]: Done

File: src/pie-chart/utils.ts [hunk 6]
```diff
@@ -163,23 +164,70 @@ export const balanceLabelNodes = (
...
 };
+
+const squareDistance = (edge: Array<number>): number => Math.pow(edge[0], 2) + Math.pow(edge[1], 2);
+
+const computeXOffset = (box: { x: number; y: number; height: number }, yOffset: number, radius: number): number => {
+  const edges = [
+    [box.x, box.y + yOffset],
+    [box.x, box.y + box.height + yOffset],
+  ];
+  const closestEdge = squareDistance(edges[0]) < squareDistance(edges[1]) ? edges[0] : edges[1];
```
[the remark this answers is not in the document]
Comment (fralongo) [reply src/pie-chart/utils.ts]: I would leave as it is as the code is more readable, and I don't think that the computation hinders performance

File: src/pie-chart/utils.ts [hunk 7]
```diff
@@ -163,23 +164,70 @@ export const balanceLabelNodes = (
...
+
+  if (squareDistance(closestEdge) < Math.pow(radius, 2)) {
+    return Math.sqrt(Math.pow(radius, 2) - Math.pow(closestEdge[1], 2)) - Math.abs(closestEdge[0]);
+  }
+  return 0;
+};
+
+export const computeSmartAngle = (startAngle: number, endAngle: number, optimize = false): number => {
+  if (!optimize || endAngle - startAngle < Math.PI / 20) {
+    return startAngle + (endAngle - startAngle) / 2;
```
[the remark this answers is not in the document]
Comment (fralongo) [reply src/pie-chart/utils.ts]: Done

File: src/pie-chart/utils.ts [hunk 8]
```diff
@@ -163,23 +164,70 @@ export const balanceLabelNodes = (
...
+  const closestEdge = squareDistance(edges[0]) < squareDistance(edges[1]) ? edges[0] : edges[1];
+
+  if (squareDistance(closestEdge) < Math.pow(radius, 2)) {
+    return Math.sqrt(Math.pow(radius, 2) - Math.pow(closestEdge[1], 2)) - Math.abs(closestEdge[0]);
+  }
+  return 0;
+};
+
+export const computeSmartAngle = (startAngle: number, endAngle: number, optimize = false): number => {
+  if (!optimize || endAngle - startAngle < Math.PI / 20) {
```
[the remark this answers is not in the document]
Comment (fralongo) [reply src/pie-chart/utils.ts]: Done

File: src/pie-chart/utils.ts (hunk 1 above)
Comment (fralongo) [reply src/pie-chart/utils.ts:194]: There are some snapshot tests for it

File: src/pie-chart/__tests__/utils.test.tsx [hunk 9]
```diff
@@ -318,3 +398,48 @@ describe.each([false, true])('getDimensionsBySize visualRefresh=%s', visualRefre
...
+
+describe('computeSmartAngle', () => {
+  const pi = Math.PI;
+  test('returns mid angle if optimization is disabled', () => {
+    expect(computeSmartAngle(0, pi / 100)).toEqual(pi / 200);
+    expect(computeSmartAngle(-1.5, 1)).toEqual(-0.25);
+    expect(computeSmartAngle(2, 4)).toEqual(3);
+    expect(computeSmartAngle(0, pi / 2)).toEqual(pi / 4);
+  });
+  test('returns mid angle if segment is too small', () => {
```
[the remark this answers is not in the document]
Comment (fralongo) [reply src/pie-chart/__tests__/utils.test.tsx:411]: Done

File: src/pie-chart/__tests__/utils.test.tsx [hunk 10]
```diff
@@ -318,3 +398,48 @@ describe.each([false, true])('getDimensionsBySize visualRefresh=%s', visualRefre
     expect(dimensions.size).toBe(matchedSize);
   });
 });
+
+describe('computeSmartAngle', () => {
```
[the remark this answers is not in the document]
Comment (fralongo) [reply src/pie-chart/__tests__/utils.test.tsx:403]: Added a test

File: src/pie-chart/labels.tsx [hunk 11]
```diff
@@ -78,7 +78,9 @@ export default <T extends PieChartProps.Datum>({
   containerRef,
 }: LabelsProps<T>) => {
   const containerBoundaries = useElementBoundaries(containerRef);
-
+  const shouldOptimizeLabels =
+    containerBoundaries.right - containerBoundaries.left - (dimensions.outerRadius + dimensions.innerLabelPadding) * 2 <
+    300;
```
Comment (rubencarvalho) [inline src/pie-chart/labels.tsx:83]: last little nit, leave it up to you if you want to change, this value could be assigned a descriptive variable, so anyone can easily identify or change this threshold.

**Review by rubencarvalho — APPROVED** [review]

**Comment by rubencarvalho** [comment]
I'll leave it up to you if you want to do it, but we could add a permutation to the pie chart where we ensure there will be truncation with the previous algorithm.

**Pull request merged**